core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -50,23 +56,36 @@ public final class SurroundingPublisher<T> implements StreamMessage<T> {
     private final StreamMessage<T> publisher;
     @Nullable
     private final T tail;
+    @Nullable
+    private final Function<@Nullable Throwable, ? extends T> finalizer;
```

Comment:
How about completely removing `tail` and using `finalizer` only?
If trailers are empty or null, we may set a special function such as 
```java
static final Function<@Nullable Throwable, ? extends T> nullTrailers = () -> null;

if (finalizer == nullTrailers) {
   // complete immediately. 
}
```

Reply:
> How about completely removing tail and using finalizer only?

aha you mean deleting constructor `new SurroundingPublisher("head", upstream, "tail")` too?

or just remove `tail` field and keep `new SurroundingPublisher("head", upstream, "tail")`? (replace "tail" to th -> "tail" finalizer internally)

Reply:
```
// T tail -> removed
Func<> finallizer
..

public onError() { // upstream error
  if (finalizer != null) {
     // If we use finalizer only, upstream error always replaced with item from finalizer.
    downstream.onNext(finalizer.apply(..));
  }
}
```
I worry that if we remove `tail` and use `finalizer` only, upstream error always replaced with finalizer.
But on our current logic with `tail`, upstream error is propagated on onError. WDYT? 🤔

Reply:
Good point. How about allowing `finalizer` not to handle the cause?
If `finalizer.apply(cause)` returns null, we can assume users do not want to recover the exception and finish the stream exceptionally.

Reply:
I applied this change myself. PTAL. 🙇‍♂️

Reply:
oh sorry for late check, thanks for your update! 🙇